tests/app/v2/notifications/test_post_letter_notifications.py

```diff
@@ -673,6 +673,28 @@ def test_post_precompiled_letter_notification_returns_201(
     assert resp_json == {'id': str(notification.id), 'reference': 'letter-reference', 'postage': expected_postage}
 
 
+def test_post_precompiled_letter_notification_if_s3_upload_fails_notification_is_not_persisted(
+    client, notify_user, mocker
+):
+    sample_service = create_service(service_permissions=['letter'])
+    s3mock = mocker.patch('app.v2.notifications.post_notifications.upload_letter_pdf', side_effect=Exception())
+    mocker.patch('app.celery.letters_pdf_tasks.notify_celery.send_task')
+    data = {
+        "reference": "letter-reference",
+        "content": "bGV0dGVyLWNvbnRlbnQ="
+    }
+
+    auth_header = create_authorization_header(service_id=sample_service.id)
+    with pytest.raises(expected_exception=Exception):
+        client.post(
+            path="v2/notifications/letter",
+            data=json.dumps(data),
+            headers=[('Content-Type', 'application/json'), auth_header])
+
```

Comment:
Optional: I think what would make this test even more rigerous would be for us to assert that `create_letter_notification` was called. So then it's very clear that we called `create_letter_notification`, then an exception was thrown but at the end there was no notification saved to the DB

Reply:
I've never mocked a method, then used side_effect=method... neat